File: src/auth/reporting.ts
/**
 * Usage reporting (Milestone 11).
 *
 * After each agent run the CLI reports usage to the backend when a valid
 * session exists. Reporting is deliberately fire-and-forget safe:
 *   - never throws (the agent keeps working no matter what),
 *   - short timeout so it cannot hang the terminal,
 *   - skips silently when the user is not logged in (pure offline mode),
 *   - clears the local session when the backend says the token is invalid.
 */
import { randomUUID } from 'node:crypto';
import type { Usage } from '../providers/types.ts';
import { ApiClient, ApiError, type UsageReportPayload } from './client.ts';
import { clearSession, loadSession, sessionExpired, type StoredSession } from './session.ts';

export interface RunReportInput {
  prompt: string;
  model: string;
  projectType: string;
  sessionId?: string;
  iterations: number;
  toolCalls: number;
  usage?: Usage;
  executionTimeMs: number;
}

export type ReportOutcome = 'reported' | 'skipped' | 'failed';

/** Build the /api/usage payload from an agent run. Returns null when there is no usage to report. */
export function buildUsageReport(input: RunReportInput): UsageReportPayload | null {
  if (!input.usage) return null;
  return {
    client_run_id: randomUUID(),
    user_id: '', // replaced by sendUsageReport with the session user's id
    session_id: input.sessionId,
    project_type: input.projectType || undefined,
    prompt: input.prompt,
    status: 'done',
    model: input.model,
    agent_turns: input.iterations,
    tool_calls: input.toolCalls,
    input_tokens: input.usage.inputTokens,
    output_tokens: input.usage.outputTokens,
    execution_time_ms: input.executionTimeMs,
  };
}

/**
 * Send a report for an already-loaded session. Never throws.
 * `clear` is the local-session removal hook (injectable for tests).
 */
export async function sendUsageReport(
  session: StoredSession,
  report: UsageReportPayload,
  clear: () => void = clearSession,
): Promise<ReportOutcome> {
  try {
    const payload: UsageReportPayload = { ...report, user_id: session.user.id };
    // Short timeout: reporting is best-effort and must never stall the CLI.
    await new ApiClient(session.apiUrl, 3000).reportUsage(session.token, payload);
    return 'reported';
  } catch (err) {
    if (err instanceof ApiError && err.status === 401) clear();
    return 'failed';
  }
}

/**
 * Report one agent run when authenticated. Returns the outcome; callers may
 * `await` it (bounded by the client timeout) or fire-and-forget.
 */
export async function reportRunUsage(
  input: RunReportInput,
  sessionOverride?: StoredSession | null,
  clear: () => void = clearSession,
): Promise<ReportOutcome> {
  const session = sessionOverride !== undefined ? sessionOverride : loadSession();
  if (!session) return 'skipped';
  if (sessionExpired(session)) {
    clear();
    return 'skipped';
  }
  const report = buildUsageReport(input);
  if (!report) return 'skipped';
  return sendUsageReport(session, report, clear);
}

File: grace/auth/reporting.py
"""Usage reporting (port of src/auth/reporting.ts).

After each agent run the CLI reports usage to the backend when a valid session
exists. Reporting is deliberately fire-and-forget safe:
  - never throws (the agent keeps working no matter what),
  - short timeout so it cannot hang the terminal,
  - skips silently when the user is not logged in (pure offline mode),
  - clears the local session when the backend says the token is invalid.
"""

import uuid

from grace.auth.client import ApiClient, ApiError
from grace.auth.session import clear_session, load_session, session_expired

# Keep in sync with the wire contract (src/api/usage.ts UsageReport).
STATUS_DONE = "done"


def build_usage_report(input_info: dict) -> dict | None:
    """Build the /api/usage payload from an agent run. None when there is no usage to report."""
    usage = input_info.get("usage")
    if not usage:
        return None
    return {
        "client_run_id": str(uuid.uuid4()),
        "user_id": "",  # replaced by send_usage_report with the session user's id
        "session_id": input_info.get("sessionId"),
        "project_type": input_info.get("projectType") or None,
        "prompt": input_info.get("prompt"),
        "status": STATUS_DONE,
        "model": input_info.get("model"),
        "agent_turns": input_info.get("iterations", 0),
        "tool_calls": input_info.get("toolCalls"),
        "input_tokens": usage.get("inputTokens", 0),
        "output_tokens": usage.get("outputTokens", 0),
        # The backend validates this as an integer (TS sends Date.now() diffs);
        # monotonic()*1000 is a float, so cast before sending.
        "execution_time_ms": int(input_info.get("executionTimeMs") or 0),
    }


def send_usage_report(session: dict, report: dict, clear=clear_session) -> str:
    """Send a report for an already-loaded session. Never throws.
    `clear` is the local-session removal hook (injectable for tests)."""
    try:
        payload = {**report, "user_id": session["user"]["id"]}
        # Short timeout: reporting is best-effort and must never stall the CLI.
        ApiClient(session["apiUrl"], 3000).report_usage(session["token"], payload)
        return "reported"
    except ApiError as err:
        if err.status == 401:
            clear()
        return "failed"
    except Exception:
        return "failed"


def report_run_usage(input_info: dict, session_override=None, clear=clear_session) -> str:
    """Report one agent run when authenticated. Returns the outcome; callers may
    await it (bounded by the client timeout) or fire-and-forget."""
    session = load_session() if session_override is None else session_override
    if not session:
        return "skipped"
    if session_expired(session):
        clear()
        return "skipped"
    report = build_usage_report(input_info)
    if not report:
        return "skipped"
    return send_usage_report(session, report, clear)
